deluge/tests/test_alertmanager.py

```diff
@@ -30,6 +55,31 @@ def handler(alert):
         self.am.register_handler('dummy_alert', handler)
         self.assertEqual(self.am.handlers['dummy_alert'], [handler])
 
+    def test_pop_alert(self):
+        def dummy_alert_1_handler(alert):
+            d.addCallback(self.assertEqual, alert.message, '1')
+            d.callback(None)
+
+        self.am.session = SessionMock()
+        self.am.register_handler('DummyAlert1', dummy_alert_1_handler)
+
+        d = Deferred()
+        return d
+
+    def test_pause_not_pop_alert(self):
+        def check_event(value):
+            event = self.am._event.isSet()
+            self.assertFalse(event)
+
+        def dummy_alert_1_handler(alert):
+            self.fail()
+
+        self.am.session = SessionMock()
+        d = component.pause(['AlertManager'])
```

Comment:
I'm worried about the timing here. I think the pause should happen and then alerts should be added to the SessionMock, then check to make sure an alert wasn't handled. The way its done right now, I think the alerts in the SessionMock could be handled before the pause occurs and the test isn't really testing anything since SessionMock.alerts == [].

Reply:
Reasonable point. Changed.

Reply:
This still doesn't address my original comments. Since you are setting self.am.session = SessionMock() before the pause happens, those alerts that are in the SessionMock.__init__ have already been processed. In other words, by the time you set the handler, there are no alerts to handle anyways so the test isn't actually testing what you expect.

What you need to do here is simulate alerts coming from libtorrent while AlertManager is paused will not trigger a handler to run, so I think SessionMock.alerts needs to be built _after_ the AlertManager is paused.

Reply:
4 years later, done 😅